### ISSUE TITLE
230717/tensorflow_회귀,분류 실습

### ISSUE THREAD

**Issue by doxgxxn** [doxgxxn/WILT#1]
회귀 -----------------------------
        
        (train_X, y_train), (test_X, y_test) = keras.datasets.boston_housing.load_data()
        
        #하이퍼파라미터 변수
        
        #학습율 => 경사하강법에서 구한 gradient(경사-미분계수)에 곱해주는 값으로 0 ~ 1 사이 실수를 지정한다. optimizer 생성때 설정.
        LEARNING_RATE = 0.001  
        N_EPOCH = 200 # 에폭수
        N_BATCH = 400 # Train 배치사이즈
        N_BATCH_VAL = 50 # Validation/evaluation 배치 사이즈. ==> 교차검증 폴더 데이터 수.
        
        
        #전처리
        
        scaler = StandardScaler()
        X_train = scaler.fit_transform(train_X)
        X_test = scaler.transform(test_X)
        
        
        #Dataset 생성
        
        train_dataset = tf.data.Dataset.from_tensor_slices((X_train, y_train)).shuffle(X_train.shape[0].batch(N_BATCH, drop_remainder=True)
        
        test_dataset = tf.data.Dataset.from_tensor_slices((X_test, y_test)).batch(N_BATCH_VAL)
        
        
        #모델 구현
        
        def get_model_boston(lr=0.01):
	        model = keras.Sequential()
	        
	        model.add(layers.Dense(32, activation='relu', input_shape(13, )))
	        model.add(layers.Dense(16, activation='relu'))
	        model.add(layers.Dense(8, activation='relu'))
        
	        model.add(layers.Dense(1))
	        
	        model.complie(optimizer=keras.optimizers.Adam(learniing_rate=lr), loss='mse'
	        return model
        
        
        model_boston = get_model_boston()
        
        #구조확인
        model_boston.summary()
        
        
        #학습
        
        hist = model_boston.fit(train_dataset, epochs=N_EPOCH, validation_data=test_dataset)
        
        
        히스토리 key값 확인
        hist.history.keys()
        
        #그래프 그리기
        plt.plot(hist.epoch, hist.history['loss'], label='Train set')
        plt.plot(hist.epoch, hist.history['val_loss'], label='Validation set')
        plt.title('에폭별 MSE loss')
        plt.ylim(0, 30)
        plt.legend()
        plt.show()
        
        
        #최종 테스트
        loss = model_boston.evaluate(test_dataset)
        
        
        #분류 ------------------------------
        
        import tesorflow as tf
        from tensorflow import keras
        from tensorflow.keras import layers, models, optimizers
        
        import numpy as np
        import matplotlib.pyplot as plt
        
        import os
        import random
        
        #seed 설정
        random.seed(0)
        np.random.seed(0)
        tf.random.set_seed(0)
        
        #후처리 class name
        class_names = np.array(['T-shirt/top', 'Trousers', 'Pulllover', 'Dress', 'Coat', 'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot'])
        
        #Data 로딩
        
        (train_image, y_train), (test_image, y_test) = keras.datasets.fashion_mnist.load_data()
        
        #validation 분리
        
        from sklearn.model_selection import train_test_split
        
        train_image, val_image, y_train, y_val = train_test_split(train_image, y_train, test_size=0.2, stratify=y_train, random_state=0)
        
        #이미지 확인
        cnt = 15 #확인할 이미지 개수
        
        plt.figure(figsize=(7,7))
        for i in range(cnt):
	        plt.subplot(5, int(cnt/5), i+1)
	        plt.imshow(train_image[i], cmap='Greys')
	        label = class_names[y_train[i]]
	        plt.title(label)
        
        plt.tight_layout()
        plt.show()
        
        #변수 설정
        LEARNING_RATE = 0.001  # 학습률
        N_EPOCH = 100          # 에폭수
        N_BATCH = 1000         # 배치 사이즈
        
        #전처리 (min/max)
        X_train = train_image.astype('float32')/255.0
        X_val = val_image.astype('float32')/255.0
        X_test = test_image.astype('float32')/255.0
        
        
        #Dataset 생성
        
        train_dataset = tf.data.Dataset.from_tensor_slices((X_train, y_train)).shuffle(X_train.shape[0]).batch(N_BATCH, drop_remainder=True)
        
        val_dataset = tf.data.Dataset.from_tensor_slices((X_val, y_val)).batch(N_BATCH)
        test_dataset = tf.data.Dataset.from_tensor_slices((X_test, y_test)).batch(N_BATCH)
        
        
        #모델 생성
        
        def get_model_fashion(lr=0.01):
	        model = keras.Sequential([
	        layers.Dense(256, activation='relu'),
	        layers.Dense(128, activation='relu'),
	        layers.Dense(64, activation='relu'),
	        layers.Dense(32, activation='relu'),
        
	        layers.Dense(10, activation='softmax', name='output'])
        
	        model.complie(optimizer=optimizzers.Adam(learning_rate=lr), loss='sparse_categorical_crossentropy', metrics=['accuracy'])
        
	        return model
        
        
        #요약 확인
        model_fashion = get_model_fashion(LEARNING_RATE)
        model_fashion.summary()
        
        #Train(학습)
        
        hist = model_fashion.fit(train_dataset, epochs=N_EPOCH, validation_data=val_dataset)
        
        
        hist.history.keys()
        hist.epoch
        
        #그래프로 확인
        
        plt. figure(figsize=(10, 5))
        plt.subplot(1,2,1)
        plt.plot(hist.epoch, hist.history['loss'], label='train')
        plt.plot(hist.epoch, hist.history['val_loss'], label='validation')
        plt.title('Loss')
        plt.xlabel('에폭수')
        plt.ylabel('loss')
        plt. legend()
        
        plt.subplot(1,2,2)
        plt.plot(his.epoch, hist.history['accuracy'], label='train')
        plt.plot(hist.epoch, hist.history['val_accuracy'], label='validation')
        plt.title('Accuracy')
        plt.xlabel('에폭수')
        plt.ylabel('accuracy')
        plt.legend()
        
        plt.tight_layout()
        plt.show()
        
        
        #최종평가
        
        loss, acc = model_fashion2.evaluate(test_dataset)
        
        print(loss, acc, sep=' , ')
        
        
        #추론
        new_X = X_test[1000:1010]
        new_X.shape
        
        pred = model_fashion.predict(new_X)
        pred.shape
        
        #추론 후처리
        label = np.argmax(pred, axis=-1)
        label_name = class_names[label]
        proba = np.max(pred, axis=-1)
        
        label, label_name, proba*100
        
        # (array([6, 3, 5, 5, 6, 4, 5, 9, 6, 3], dtype=int64),
         array(['Shirt', 'Dress', 'Sandal', 'Sandal', 'Shirt', 'Coat', 'Sandal',
                'Ankle boot', 'Shirt', 'Dress'], dtype='<U11'),
         array([ 99.685356,  99.99996 , 100.      , 100.      ,  73.670044,
                 95.15789 , 100.      , 100.      ,  99.99991 ,  99.999466],
               dtype=float32))
        
        
        for idx, (n, p) in enumerate(zip(label_name, proba), start=1):
	        print(f'{idx}. {n} - {int(p*100)}%')
